modules/apps/site/site-cms-site-initializer/src/main/java/com/liferay/site/cms/site/initializer/internal/display/context/SpacesSectionDisplayContext.java

```diff
@@ -0,0 +1,103 @@
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.site.cms.site.initializer.internal.display.context;
+
+import com.liferay.headless.asset.library.dto.v1_0.AssetLibrary;
+import com.liferay.headless.asset.library.resource.v1_0.AssetLibraryResource;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.portal.vulcan.pagination.Page;
+import com.liferay.portal.vulcan.pagination.Pagination;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * @author Roberto Díaz
+ */
+public class SpacesSectionDisplayContext {
+
+	public SpacesSectionDisplayContext(
+		HttpServletRequest httpServletRequest,
+		AssetLibraryResource.Factory assetLibraryResourceFactory) {
+
+		_assetLibraryResourceFactory = assetLibraryResourceFactory;
+
+		_themeDisplay = (ThemeDisplay)httpServletRequest.getAttribute(
+			WebKeys.THEME_DISPLAY);
+	}
+
+	public Page<AssetLibrary> getPage() throws Exception {
+		AssetLibraryResource.Builder builder =
+			_assetLibraryResourceFactory.create();
+
+		AssetLibraryResource assetLibraryResource = builder.user(
+			_themeDisplay.getUser()
+		).build();
+
+		Page<AssetLibrary> assetLibrariesPinnedByMePage =
+			assetLibraryResource.getAssetLibrariesPinnedByMePage();
+
+		Page<AssetLibrary> assetLibrariesPage =
+			assetLibraryResource.getAssetLibrariesPage(
+				null, null, null,
+				Pagination.of(
+					1, (int)(5 + assetLibrariesPinnedByMePage.getPageSize())),
+				null);
+
+		return Page.of(
+			assetLibrariesPage.getActions(),
+			_getItems(assetLibrariesPinnedByMePage, assetLibrariesPage),
+			Pagination.of(1, 5), assetLibrariesPage.getTotalCount());
+	}
+
+	private Collection<AssetLibrary> _getItems(
+		Page<AssetLibrary> assetLibrariesPinnedByMePage,
+		Page<AssetLibrary> assetLibrariesPage) {
+
+		List<AssetLibrary> assetLibrariesPinnedByMePageItems = new ArrayList<>(
+			assetLibrariesPinnedByMePage.getItems());
+
+		if (assetLibrariesPinnedByMePageItems.size() == 5) {
+			return assetLibrariesPinnedByMePageItems;
+		}
+
+		List<AssetLibrary> assetLibrariesPageItems = new ArrayList<>(
+			assetLibrariesPage.getItems());
+
+		if (assetLibrariesPinnedByMePageItems.isEmpty()) {
+			if (assetLibrariesPageItems.isEmpty()) {
+				Collections.emptyList();
+			}
+			else if (assetLibrariesPageItems.size() <= 5) {
+				return assetLibrariesPageItems;
+			}
+			else if (assetLibrariesPageItems.size() > 5) {
+				return assetLibrariesPageItems.subList(0, 5);
+			}
+		}
+
+		List<AssetLibrary> items = assetLibrariesPinnedByMePageItems;
+
+		for (AssetLibrary assetLibraryItem : assetLibrariesPageItems) {
+			items.add(assetLibraryItem);
+
+			if (items.size() == 5) {
+				return items;
+			}
+		}
+
+		return assetLibrariesPageItems;
+	}
```

Comment:
I think this is much more complicated than it needs to be. Not sure if I'm missing something, but why not:
```
if (pinnedALs.size() >= 5) {
    return pinnedALs.subList(0, 5);
}

List<AssetLibrary> padding = _getPadding(); // request 5 items, no more are needed;

padding.removeAll(pinnedALs);

pinnedALs.addAll(padding);

return pinnedALs.subList(0, 5);
```